judge/judgedaemon.main.php

```diff
@@ -419,12 +419,15 @@ function fetch_executable_internal(
                         $buildscript .= "javac -cp ./ -d ./ '$source'.java\n";
                         $buildscript .= "echo '#!/bin/sh' > run\n";
                         // no main class detection here
-                        $buildscript .= "echo 'java -cp ./ '$source >> run\n";
+                        $buildscript .= "echo COMPARE_DIR=\$(dirname \"\$0\") >> run\n";
+                        $buildscript .= "echo 'java -cp \"\$COMPARE_DIR\" '$source \"\$@\" >> run\n";
```

Comment:
The single quotes look strange here: they don't wrap around the full string to be echo-ed and it seems that this could cause issues if `$source` contains weird characters, e.g. spaces/tabs. I think the simplest solution would be go apply `dj_escapeshellarg` to `$source`. Then it can be used as is here, but in the other commands, the quotes around it should be removed.

Reply:
Also, I think that `\"\$@\"` is not correctly quoted, as the run script must contain `"$@"`. Can you check that that is the case?

Reply:
I will check. The single quote was already weird before right?

Reply:
Indeed, it was there already.

Reply:
They were wrong indeed, now they seem to be correct. This is the contents of `build` and `run`:

```bash
domjudge@domjudge-contributor:/Users/nicky/Projects/domjudge/output/judgings/domjudge-contributor-1/endpoint-default/executable/compare/31/67bf2c1b54f229b56dce724588d6bedf/build$ cat build
#!/bin/sh

javac -cp ./ -d ./ 'Main.java'
echo '#!/bin/sh' > run
echo 'COMPARE_DIR=$(dirname "$0")' >> run
echo 'java -cp "$COMPARE_DIR" Main "$@"' >> run
chmod +x run


domjudge@domjudge-contributor:/Users/nicky/Projects/domjudge/output/judgings/domjudge-contributor-1/endpoint-default/executable/compare/31/67bf2c1b54f229b56dce724588d6bedf/build$ cat run
#!/bin/sh
COMPARE_DIR=$(dirname "$0")
java -cp "$COMPARE_DIR" Main "$@"
```